### DISCUSSION TITLE
Get bounds, center, and layer names programmatically

### DISCUSSION THREAD

**Discussion by giswqs** [protomaps/PMTiles#255] [Q&A] [answered]
What's the recommended way to get a PMTiles archive's bounds, center, and layer names programmtically besides using `pmtiles show`? I am using folium to render PMTiles, and I need to automatically center the map on the map based on the bounding box, similar to the PMTiles View. I also need to retrieve the layer names programmatically so that I can set layer styles accordingly. 

![image](https://github.com/protomaps/PMTiles/assets/5016453/9173c1a2-4940-41a9-b9c2-a9707b5f0398)

**Comment by bdon** [comment]
what language do you need to do this in?

An example of getting `maxZoom` in JavaScript is here: https://github.com/protomaps/PMTiles/blob/main/js/examples/leaflet.html#L21C23-L21C23

`center` and `bounds` are also on the header object. You need to block your load on getting the header first before initializing the map.

For vector-specific archives, you also need to block on loading the JSON metadata which contains the `vector_layers` key.

**Reply by giswqs** [reply]
I am using Python.

**Reply by bdon** [reply]
Do you have an example of where you set the bounds, center, zoom for a non-PMTiles data source in Python?

**Reply by giswqs** [reply]
Set center and fit bounds are usually applied to the map, not to a data source. What I want is a Python equivalent to your JS PMTiles Viewer. When a PMTiles archive is loaded, it can automatically zoom to the bounds of the layer. For now, I have to use `pmtiles show` to inspect the layer center manually and then set that in `folium.Map(location=[latitude, longitude])` when initializing the map. 

For folium:
https://python-visualization.github.io/folium/latest/reference.html
![image](https://github.com/protomaps/PMTiles/assets/5016453/818802df-88a3-487f-977c-ed8108d97288)

For leafmap:
https://leafmap.org/notebooks/82_pmtiles/

**Reply by bdon** [reply]
The Python equivalent is here: https://github.com/protomaps/PMTiles/blob/main/python/pmtiles/reader.py#L34 but please be aware this python package is scarcely used and poorly documented - let me know if that looks like a viable approach

**Reply by giswqs** [reply]
I installed pmtiles using `pip install pmtiles` and then was able to read the header and metadata, but the results are different from using the pmtiles binary command `pmtiles show`. Also, it the seems Python API can only inspect local PMTiles while the PMTiles command can handle both local and remote files. What I need is a way to retrieve the bounds, center, and layer names from a remote PMTiles using Python. Is this alaready available through the PMTiles Python library?

```python
from pmtiles.reader import Reader, MmapSource
with open('protomaps(vector)ODbL_firenze.pmtiles', 'rb') as f:
    reader = Reader(MmapSource(f))
    print(reader.header())
    print(reader.metadata())
```

> {'version': 3, 'root_offset': 127, 'root_length': 406, 'metadata_offset': 533, 'metadata_length': 575, 'leaf_directory_offset': 1108, 'leaf_directory_length': 0, 'tile_data_offset': 1108, 'tile_data_length': 3958871, 'addressed_tiles_count': 108, 'tile_entries_count': 108, 'tile_contents_count': 106, 'clustered': True, 'internal_compression': <Compression.GZIP: 2>, 'tile_compression': <Compression.GZIP: 2>, 'tile_type': <TileType.MVT: 1>, 'min_zoom': 0, 'max_zoom': 14, 'min_lon_e7': 111540260, 'min_lat_e7': 437270125, 'max_lon_e7': 113289395, 'max_lat_e7': 438325455, 'center_zoom': 0, 'center_lon_e7': 112414827, 'center_lat_e7': 437797790}
{'attribution': '<a href="https://protomaps.com" target="_blank">Protomaps</a> © <a href="https://www.openstreetmap.org" target="_blank"> OpenStreetMap</a>', 'name': 'protomaps 2023-01-18T07:49:39Z', 'type': 'baselayer', 'vector_layers': [{'fields': {}, 'id': 'earth'}, {'fields': {'boundary': 'string', 'landuse': 'string', 'leisure': 'string', 'name': 'string', 'natural': 'string'}, 'id': 'natural'}, {'fields': {'aeroway': 'string', 'amenity': 'string', 'area:aeroway': 'string', 'highway': 'string', 'landuse': 'string', 'leisure': 'string', 'man_made': 'string', 'name': 'string', 'place': 'string', 'pmap:kind': 'string', 'railway': 'string', 'sport': 'string'}, 'id': 'land'}, {'fields': {'landuse': 'string', 'leisure': 'string', 'name': 'string', 'natural': 'string', 'water': 'string', 'waterway': 'string'}, 'id': 'water'}, {'fields': {'natural': 'string', 'waterway': 'string'}, 'id': 'physical_line'}, {'fields': {'building:part': 'string', 'height': 'number', 'layer': 'string', 'name': 'string'}, 'id': 'buildings'}, {'fields': {'ele': 'number', 'name': 'string', 'natural': 'string', 'place': 'string'}, 'id': 'physical_point'}, {'fields': {'capital': 'string', 'country_code_iso3166_1_alpha_2': 'string', 'name': 'string', 'place': 'string', 'pmap:kind': 'string', 'pmap:rank': 'string', 'population': 'string'}, 'id': 'places'}, {'fields': {'bridge': 'string', 'highway': 'string', 'layer': 'string', 'oneway': 'string', 'pmap:kind': 'string', 'ref': 'string', 'tunnel': 'string'}, 'id': 'roads'}, {'fields': {'aerialway': 'string', 'aeroway': 'string', 'highspeed': 'string', 'layer': 'string', 'name': 'string', 'network': 'string', 'pmap:kind': 'string', 'railway': 'string', 'ref': 'string', 'route': 'string', 'service': 'string'}, 'id': 'transit'}, {'fields': {'amenity': 'string', 'cuisine': 'string', 'name': 'string', 'railway': 'string', 'religion': 'string', 'shop': 'string', 'tourism': 'string'}, 'id': 'pois'}, {'fields': {'pmap:min_admin_level': 'number'}, 'id': 'boundaries'}, {'fields': {}, 'id': 'mask'}]}

![image](https://github.com/protomaps/PMTiles/assets/5016453/d35610e9-3070-4fed-a603-f988e4869fd9)

**Reply by bdon** [reply]
Those results are the same, just the `e7` means it is multiplied by 10000000.

It should be quite simple to add remote file support by implementing your own `Source` class with a `get_bytes` method. The `pmtiles` library does not provide one right now for remote sources, because different applications may prefer to use `urllib3`, `requests`, etc. Is there an existing HTTP library you're using?

**Comment by giswqs** [comment, marked as answer]
I was able to implement the following two functions for getting PMTiles header and metadata. It works with both local and remote PMTiles.  https://github.com/opengeos/leafmap/pull/566

```python
def pmtiles_header(input_file: str):
    """
    Fetch the header information from a local or remote .pmtiles file.

    This function retrieves the header from a PMTiles file, either local or hosted remotely.
    It deserializes the header and calculates the center and bounds of the tiles from the
    given metadata in the header.

    Args:
        input_file (str): Path to the .pmtiles file, or its URL if the file is hosted remotely.

    Returns:
        dict: A dictionary containing the header information, including center and bounds.

    Raises:
        ImportError: If the pmtiles library is not installed.
        ValueError: If the input file is not a .pmtiles file or if it does not exist.

    Example:
        >>> header = pmtiles_header("https://example.com/path/to/tiles.pmtiles")
        >>> print(header["center"])
        [52.5200, 13.4050]

    Note:
        If fetching a remote PMTiles file, this function only downloads the first 127 bytes
        of the file to retrieve the header.
    """

    import requests

    try:
        from pmtiles.reader import Reader, MmapSource
        from pmtiles.tile import deserialize_header
    except ImportError:
        print(
            "pmtiles is not installed. Please install it using `pip install pmtiles`."
        )
        return

    if not input_file.endswith(".pmtiles"):
        raise ValueError("Input file must be a .pmtiles file.")

    if input_file.startswith("http"):
        # Fetch only the first 127 bytes
        headers = {"Range": "bytes=0-127"}
        response = requests.get(input_file, headers=headers)
        header = deserialize_header(response.content)

    else:
        if not os.path.exists(input_file):
            raise ValueError(f"Input file {input_file} does not exist.")

        with open(input_file, "rb") as f:
            reader = Reader(MmapSource(f))
            header = reader.header()

    header["center"] = [header["center_lat_e7"] / 1e7, header["center_lon_e7"] / 1e7]
    header["bounds"] = [
        header["min_lon_e7"] / 1e7,
        header["min_lat_e7"] / 1e7,
        header["max_lon_e7"] / 1e7,
        header["max_lat_e7"] / 1e7,
    ]

    return header


def pmtiles_metadata(input_file: str) -> Dict[str, Union[str, int, List[str]]]:
    """
    Fetch the metadata from a local or remote .pmtiles file.

    This function retrieves metadata from a PMTiles file, whether it's local or hosted remotely.
    If it's remote, the function fetches the header to determine the range of bytes to download
    for obtaining the metadata. It then reads the metadata and extracts the layer names.

    Args:
        input_file (str): Path to the .pmtiles file, or its URL if the file is hosted remotely.

    Returns:
        dict: A dictionary containing the metadata information, including layer names.

    Raises:
        ImportError: If the pmtiles library is not installed.
        ValueError: If the input file is not a .pmtiles file or if it does not exist.

    Example:
        >>> metadata = pmtiles_metadata("https://example.com/path/to/tiles.pmtiles")
        >>> print(metadata["layer_names"])
        ['buildings', 'roads']

    Note:
        If fetching a remote PMTiles file, this function may perform multiple requests to minimize
        the amount of data downloaded.
    """

    import json
    import requests

    try:
        from pmtiles.reader import Reader, MmapSource, MemorySource
    except ImportError:
        print(
            "pmtiles is not installed. Please install it using `pip install pmtiles`."
        )
        return

    if not input_file.endswith(".pmtiles"):
        raise ValueError("Input file must be a .pmtiles file.")

    header = pmtiles_header(input_file)
    metadata_offset = header["metadata_offset"]
    metadata_length = header["metadata_length"]

    if input_file.startswith("http"):
        headers = {"Range": f"bytes=0-{metadata_offset + metadata_length}"}
        response = requests.get(input_file, headers=headers)
        content = MemorySource(response.content)
        metadata = Reader(content).metadata()
    else:
        with open(input_file, "rb") as f:
            reader = Reader(MmapSource(f))
            metadata = reader.metadata()
            metadata["vector_layers"] = json.loads(metadata["json"])["vector_layers"]

    layers = metadata["vector_layers"]
    layer_names = [layer["id"] for layer in layers]
    metadata["layer_names"] = layer_names
    metadata["center"] = header["center"]
    metadata["bounds"] = header["bounds"]
    return metadata
```

Examples:
```python
import leafmap

url = "https://storage.googleapis.com/ahp-research/overture/pmtiles/overture.pmtiles"
metadata = leafmap.pmtiles_metadata(url)
print(metadata['layer_names'])
print(metadata['bounds'])
print(metadata['center'])
```

> ['admins', 'buildings', 'places', 'roads']
[3.295898, 50.746884, 7.245483, 53.5762]
[52.963529, 4.776306]
